en/controllers/middleware.rst

```diff
@@ -0,0 +1,252 @@
+
+After the middleware stack has been prepared by the application, the
+``Server.buildMiddleware`` event is triggered. This event can be useful to add
+middleware from plugins. Plugins can register listeners in their bootstrap
+scripts, that add middleware::
+
+    // In ContactManager plugin bootstrap.php
+    use Cake\Event\EventManager;
+
+    EventManager::instance()->on(
+        'Server.buildMiddleware',
+        function ($event, $middleware) {
+            $middleware->push(new PluginMiddleware());
+        });
+
+PSR7 Requests and Responses
+===========================
+
+Middleware and the new HTTP stack are built on top of the `PSR7 Request
+& Response Interfaces <http://www.php-fig.org/psr/psr-7/>`__. While all
+middleware will be exposed to these interfaces, your controllers, components,
+and views will *not*.
+
+Interacting with Requests
+-------------------------
+
+The ``RequestInterface`` provides methods for interacting with the headers,
+method, URI, and body of a request. To interact with the headers, you can::
+
+    // Read a header as text
+    $value = $request->getHeaderLine(‘Content-Type’);
+
+    // Read header as an array
+    $value = $request->getHeader(‘Content-Type’);
+
+    // Read all the headers as an associative array.
+    $headers = $request->getHeaders();
+
+Requests also give access to the cookies and uploaded files they contain::
+
+    // Get an array of cookie values.
+    $cookies = $request->getCookieParams();
+
+    // Get a list of UploadedFile objects
+    $files = $request->getUploadedFiles();
+
+    // Read the file data.
+    $files[0]->getStream();
+    $files[0]->getSize();
+    $files[0]->getClientFileName();
+
+    // Move the file.
+    $files[0]->moveTo($targetPath);
+
+Requests contain a URI object, which contains methods for interacting with the
+requested URI::
+
+    // Get the URI
+    $uri = $request->getUri();
+
+    // Read data out of the URI.
+    $path = $uri->getPath();
+    $query = $uri->getQuery();
+    $host = $uri->getHost();
+
+Lastly, you can interact with a request's 'attributes'. CakePHP uses these
+attributes to carry framework specific request parameters. There are 3 important
+attributes in any request handled by CakePHP:
+
+* ``base`` contains the base directory for your application if there is one.
+* ``webroot`` contains the webroot directory for your application.
+* ``params`` contains the results of route matching once routing rules have been
+  processed.
+
+Interacting with Responses
+--------------------------
+
+The methods available to create a server response are the same as those
+available when interacting with :ref:`httpclient-response-objects`. While the
+interface is the same the usage scenarios are different.
+
+When modifying the response, it is important to remember that responses are
+**immutable**. You must always remember to store the results of any setter
+method. For example::
+
+    // This does *not* change the response. As our new object
```

Comment:
"This does *not* change the response, as the resulting object, containing the changes, is not stored."?